README.md
# crypto-exchanges-ws-client

Node.js implementation of websocket protocol used by [Crypto Exchange Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway)

## What it does

* Implement methods for subscribing to tickers, order books, trade, klines for Bittrex, Poloniex, Binance, Kucoin & OKex (all exchanges supported by [Crypto Exchange Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway))

* Implement methods for subscribing to klines

* Implement method to subscribe to Ticker Monitor (custom price alerts) feed

* Handle automatic reconnection in (I think !) every possible scenario

## Installation

```
npm install crypto-exchanges-ws-client
```

## How to use it

See [documentation in _doc_ directory](https://github.com/aloysius-pgast/crypto-exchanges-ws-client-nodejs/tree/master/doc/) for a description of supported API

See [examples in _examples_ directory](https://github.com/aloysius-pgast/crypto-exchanges-ws-client-nodejs/tree/master/examples/) for an overview of what this library can do

## Code sample

Subscribe to `4h` & `1d` klines, for `USDT-BTC` & `USDT-ETH` pairs on `Binance` & `Kucoin` exchanges

Example assume that [Crypto Exchange Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway) is running locally on ports *8000* & *8001*

```
"use strict";
const WebSocket = require('crypto-exchanges-ws-client');

const exchanges = ['binance', 'kucoin'];
const pairs = ['USDT-BTC', 'USDT-ETH'];
const klinesInterval = ['4h', '1d'];

const uri = 'ws://127.0.0.1:8001';
const client = new WebSocket(uri);

client.on('connected', (data) => {
    console.log(`We connected/reconnected : ${JSON.stringify(data)}`);
});

// connection to gateway was lost (reconnection will be automatic)
client.on('disconnected', (data) => {
    console.log(`We've been disconnected : ${JSON.stringify(data)}`);
});

client.on('ready', (data) => {
    console.log(`Ready to subscribe : ${JSON.stringify(data)}`);
    for (let exchange of exchanges) {
        for (let pair of pairs) {
            for (let interval of klinesInterval) {
                client.subscribeToKlines(exchange, pairs, interval);
            }
        }
    }
});

client.on('kline', (evt) => {
    console.log(JSON.stringify(evt));
});
```

Output should be similar to

```
We connected/reconnected : {"connectionId":1}                                                                                                                                                                                                
Ready to subscribe : {"sessionId":"rpc.bdff7468-5c7d-412f-a44a-4b37dab2ecc8","isNew":true}                                                                                                                                                   
{"exchange":"kucoin","pair":"USDT-BTC","interval":"1d","data":{"timestamp":1567987200,"open":10384.3,"high":10403.1,"low":10242.5,"close":10268.7,"volume":646.80699337,"remainingTime":61621,"closed":false}}                               
{"exchange":"kucoin","pair":"USDT-ETH","interval":"1d","data":{"timestamp":1567987200,"open":181.29,"high":181.6,"low":177,"close":177.44,"volume":20139.11233964,"remainingTime":61621,"closed":false}}                                     
{"exchange":"kucoin","pair":"USDT-BTC","interval":"1d","data":{"timestamp":1567987200,"open":10384.3,"high":10403.1,"low":10242.5,"close":10271.2,"volume":646.93397568,"remainingTime":61621,"closed":false}}                               
{"exchange":"kucoin","pair":"USDT-ETH","interval":"1d","data":{"timestamp":1567987200,"open":181.29,"high":181.6,"low":177,"close":177.44,"volume":20139.11667794,"remainingTime":61621,"closed":false}}                                     
{"pair":"USDT-BTC","interval":"4h","data":{"timestamp":1568001600,"open":10300.22,"close":10271.88,"high":10330.33,"low":10250.01,"volume":4470.045063,"remainingTime":4017,"closed":false},"exchange":"binance"}                            
{"pair":"USDT-BTC","interval":"1d","data":{"timestamp":1567987200,"open":10381.24,"close":10271.88,"high":10404.74,"low":10250.01,"volume":8118.81537,"remainingTime":61617,"closed":false},"exchange":"binance"}                            
{"pair":"USDT-ETH","interval":"1d","data":{"timestamp":1567987200,"open":181.18,"close":177.64,"high":181.65,"low":177,"volume":70992.51618,"remainingTime":61617,"closed":false},"exchange":"binance"}
{"pair":"USDT-ETH","interval":"4h","data":{"timestamp":1568001600,"open":178.25,"close":177.64,"high":179.01,"low":177,"volume":29569.11772,"remainingTime":4017,"closed":false},"exchange":"binance"}
{"pair":"USDT-BTC","interval":"4h","data":{"timestamp":1568001600,"open":10300.22,"close":10275.1,"high":10330.33,"low":10250.01,"volume":4470.410513,"remainingTime":4015,"closed":false},"exchange":"binance"}
{"pair":"USDT-BTC","interval":"1d","data":{"timestamp":1567987200,"open":10381.24,"close":10275.1,"high":10404.74,"low":10250.01,"volume":8119.18082,"remainingTime":61615,"closed":false},"exchange":"binance"}
{"pair":"USDT-ETH","interval":"1d","data":{"timestamp":1567987200,"open":181.18,"close":177.64,"high":181.65,"low":177,"volume":71009.36855,"remainingTime":61615,"closed":false},"exchange":"binance"}
{"pair":"USDT-ETH","interval":"4h","data":{"timestamp":1568001600,"open":178.25,"close":177.64,"high":179.01,"low":177,"volume":29585.97009,"remainingTime":4015,"closed":false},"exchange":"binance"}
{"pair":"USDT-BTC","interval":"4h","data":{"timestamp":1568001600,"open":10300.22,"close":10275.2,"high":10330.33,"low":10250.01,"volume":4471.59265,"remainingTime":4013,"closed":false},"exchange":"binance"}
{"pair":"USDT-BTC","interval":"1d","data":{"timestamp":1567987200,"open":10381.24,"close":10275.2,"high":10404.74,"low":10250.01,"volume":8120.362957,"remainingTime":61613,"closed":false},"exchange":"binance"}
{"pair":"USDT-ETH","interval":"1d","data":{"timestamp":1567987200,"open":181.18,"close":177.63,"high":181.65,"low":177,"volume":71009.43655,"remainingTime":61612,"closed":false},"exchange":"binance"}
{"pair":"USDT-ETH","interval":"4h","data":{"timestamp":1568001600,"open":178.25,"close":177.63,"high":179.01,"low":177,"volume":29586.03809,"remainingTime":4012,"closed":false},"exchange":"binance"}
```


CHANGELOG.md
# Change Log

## [v1.0.6]
* Add simple example in README
* Ensure exceptions in client event handlers are not silently caught by module
* Update various dependencies

## [v1.0.5]
* Query parameters _expires_ & _timeout_ were only sent when _sid_ was provided in url
* Ensure that a new _sid_ is generated by gateway on reconnect, when using ephemeral sessions

## [v1.0.4]
* Support for Ticker Monitor (requires [Crypto Exchange Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway) >= 1.6.0)

## [v1.0.3]
* Support for Klines (requires [Crypto Exchange Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway) >= 1.4.0)

## [v1.0.2]
* Allows to provide _API Key_ in constructor (in case _API Key_ is configured on gateway)

## [v1.0.1]
* New methods to cancel all subscriptions for a single exchange or globally

## [v1.0.0]
* Initial version


lib/websocket-connection.js
"use strict";
const retry = require('retry');
const WebSocket = require('ws');
const EventEmitter = require('events');
const debug = require('debug')('CEWSC:WebSocketConnection');

// how long should we wait before trying to reconnect upon connection failure
const RETRY_DELAY = 10 * 1000;

// infinite retry
const RETRY_COUNT = -1;

// connection will be closed if we don't receive pong after timeout
const PING_TIMEOUT = 30000;

// how long do we want to wait for handshake
const HANDSHAKE_TIMEOUT = 10 * 1000;

// default user agent
const USER_AGENT = 'CEWSC 1.0';

// connection states
const STATE_NEW = 0;
const STATE_CONNECTING = 1;
const STATE_CONNECTED = 2;
const STATE_DISCONNECTING = 3;
const STATE_DISCONNECTED = 3;

/*
   Following events can be emitted

   1) message, when a message is received

   Data will contain the message received

   2) connectionError, when a connection error occurs (ie: WS cannot be connnected)

   Data will be an object {attempts:integer,retry:boolean,error:err}

   - attempts : number of attempts to connect
   - retry : whether or not there are retry left
   - error : the connection error which occurred

   3) disconnected, when websocket connection has been disconnected

   This is a final event. Websocket won't be reconnected. A new WebsocketConnection object should be used
   Event will not be emitted in case connection is disconnected by client or on connection failure
   Event will not be emitted for connection/reconnection error (event connectionError will be emitted instead)

   Data will be an object {code:integer,reason:string}

   4) connected, when websocket connection is ready to receive message

   Event will only be emitted once in the lifetime of the object

 */

class WebSocketConnection extends EventEmitter
{

constructor(uri, options)
{
    super();
    this._uri = uri;
    this._retryCount = RETRY_COUNT;
    this._retryDelay = RETRY_DELAY;
    this._pingTimeout = PING_TIMEOUT;
    this._apiKey = '';
    if (undefined !== options)
    {
        // retry count
        if (undefined !== options.retryCount)
        {
            this._retryCount = options.retryCount;
        }
        if (undefined !== options.retryDelay)
        {
            this._retryDelay = options.retryDelay;
        }
        if (undefined !== options.pingTimeout)
        {
            this._pingTimeout = options.pingTimeout;
        }
        if (undefined !== options.apiKey)
        {
            this._apiKey = options.apiKey;
        }
    }
    this._ws = null;
    // when WS successfully connected
    this._timestamp = null;
    this._connectionState = STATE_NEW;
    this._ignoreCloseEvent = true;
}

isConnected()
{
    return STATE_CONNECTED == this._connectionState;
}

send(data)
{
    if (STATE_CONNECTED != this._connectionState)
    {
        return false;
    }
    this._ws.send(data);
}

disconnect()
{
    if (STATE_DISCONNECTED == this._connectionState || STATE_DISCONNECTING == this._connectionState)
    {
        return;
    }
    this._connectionState = STATE_DISCONNECTING;
    this._finalize(false, STATE_DISCONNECTED);
    return;
}

connect()
{
    if (STATE_NEW !== this._connectionState)
    {
        return false;
    }
    let attempt = 1;
    this._connectionState = STATE_CONNECTING;
    let self = this;
    try
    {
        let retryOptions = {
            minTimeout:this._retryDelay,
            // do not use any exponential factor
            factor:1,
            randomize:false
        };
        if (-1 == this._retryCount)
        {
            retryOptions.forever = true;
        }
        else
        {
            retryOptions.retries = this._retryCount;
        }
        let wsOptions = {
            perMessageDeflate: false,
            handshakeTimeout:HANDSHAKE_TIMEOUT,
            headers: {
                'User-Agent': USER_AGENT
            }
        }
        if ('' != this._apiKey)
        {
            wsOptions.headers['ApiKey'] = this._apiKey;
        }
        let operation = retry.operation(retryOptions);
        operation.attempt(function(currentAttempt){
            // connection has already been disconnected by client
            if (STATE_CONNECTING != self._connectionState)
            {
                return;
            }
            attempt = currentAttempt;
            let doRetry = true;
            let ws = new WebSocket(self._uri, wsOptions);
            let ignoreErrorEvent = false;
            let skipCloseEvent = false;
            ws.on('open', function() {
                // connection has already been disconnected by client
                if (STATE_CONNECTING != self._connectionState)
                {
                    return;
                }
                self._connectionState = STATE_CONNECTED;
                if (debug.enabled)
                {
                    debug("WS is connected");
                }
                self._ignoreCloseEvent = false;
                skipCloseEvent = false;
                self._timestamp = new Date().getTime();
                self._ws = this;
                // start ping/pong
                if (0 != self._pingTimeout)
                {
                    let _ws = this;
                    _ws.isAlive = false;
                    // initial ping
                    _ws.ping('', true, true);
                    let interval = setInterval(function() {
                        if (WebSocket.OPEN != _ws.readyState)
                        {
                            clearTimeout(interval);
                            return;
                        }
                        if (!_ws.isAlive)
                        {
                            if (debug.enabled)
                            {
                                debug("WS timeout : timeout = %d", self._pingTimeout);
                            }
                            _ws.terminate();
                            clearTimeout(interval);
                            return;
                        }
                        _ws.isAlive = false;
                        _ws.ping('', true, true);
                    }, self._pingTimeout);
                }
                self.emit('connected');
            });
            ws.on('message', function(message) {
                self.emit('message', message);
            });
            ws.on('error', function(e) {
                if (ignoreErrorEvent)
                {
                    return;
                }
                // connection has already been disconnected by client
                if (STATE_CONNECTING != self._connectionState)
                {
                    return;
                }
                let err = {code:e.code,message:e.message}
                if (debug.enabled)
                {
                    debug("WS error (attempt %d/%s) : %s", attempt, -1 === self._retryCount ? 'unlimited' : (1 + self._retryCount), JSON.stringify(err));
                }
                skipCloseEvent = true;
                self._ws = null;
                this.terminate();
                // ws is not open yet, likely to be a connection error
                if (null === self._timestamp)
                {
                    if (doRetry && operation.retry(err))
                    {
                        self.emit('connectionError', {attempts:attempt,retry:true,error:err});
                        return;
                    }
                    self.emit('connectionError', {attempts:attempt,retry:false,error:err});
                }
            });
            // likely to be an auth error
            ws.on('unexpected-response', function(request, response){
                // connection has already been disconnected by client
                if (STATE_CONNECTING != self._connectionState)
                {
                    return;
                }
                let err = {code:response.statusCode,message:response.statusMessage};
                ignoreErrorEvent = true;
                skipCloseEvent = true;
                if (debug.enabled)
                {
                    debug("WS unexpected-response (attempt %d/%s) : %s", self._uri, attempt, -1 === self._retryCount ? 'unlimited' : (1 + self._retryCount), JSON.stringify(err));
                }
                self._ws = null;
                if (doRetry && operation.retry(err))
                {
                    self.emit('connectionError', {attempts:attempt,retry:true,error:err});
                    return;
                }
                self.emit('connectionError', {attempts:attempt,retry:false,error:err});
            });
            ws.on('close', function(code, reason){
                if (self._ignoreCloseEvent)
                {
                    return;
                }
                // connection has already been disconnected by client
                if (STATE_CONNECTING != self._connectionState && STATE_CONNECTED != self._connectionState)
                {
                    return;
                }
                if (debug.enabled)
                {
                    debug("WS closed : code = %d, reason = '%s'", code, reason);
                }
                self._ws = null;
                self._finalize(true, STATE_DISCONNECTED);
                if (!skipCloseEvent)
                {
                    self.emit('disconnected', {code:code, reason:reason});
                }
            });
            // reply to ping
            ws.on('ping', function(data){
                this.pong('', true, true);
            });
            ws.on('pong', function(data){
                this.isAlive = true;
            });
        });
    }
    catch (e)
    {
        throw e;
    }
    return true;
}

/**
 * Used to do a bit of cleaning (close ws, abort ...)
 *
 * @param {boolean} terminate indicates whether or not WS should be terminated vs closed
 * @param {integer} newState new connection state
 */
_finalize(terminate, newState)
{
    // close ws
    if (null !== this._ws)
    {
        let ws = this._ws;
        this._ws = null;
        this._ignoreCloseEvent = true;
        try
        {
            if (terminate)
            {
                ws.terminate();
            }
            else
            {
                ws.close();
            }
        }
        catch (e)
        {
            // do nothing
        }
    }
    this._connectionState = newState;
}

}

module.exports = WebSocketConnection;


examples/subscribe_to_all_tickers.js
"use strict";
const WebSocketClient = require('../lib/client');

/*
 * Below example assumes that your gateway is listening for WS on 127.0.0.1:8001
 */

const uri = 'ws://127.0.0.1:8001'
const client = new WebSocketClient(uri);

//-- Connection event handlers
// socket is connected but not ready to receive subscription yet
client.on('connected', function(data){
    console.log(`We connected/reconnected : ${JSON.stringify(data)}`);
});
// connection to gateway was lost (reconnection will be automatic)
client.on('disconnected', function(data){
    console.log(`We've been disconnected : ${JSON.stringify(data)}`);
});
// temporary connection error, retry will be automatic
client.on('connectionError', function(data){
    console.log(`Connection error : ${JSON.stringify(data)}`);
});
// this is were we can start subscriptions
client.on('ready', function(data){
    console.log(`Ready to subscribe : ${JSON.stringify(data)}`);

    // subscribe to all ticker pairs having NEO as currency on Bittrex
    client.getPairs('bittrex', {currency:'NEO'}, function(result, error){
        if (null !== error)
        {
            console.error(`Could not retrieve pairs : ${JSON.stringify(error)}`);
            client.disconnect();
            process.exit(1);
            return;
        }
        client.subscribeToTickers('bittrex', Object.keys(result));
    });
});

//-- Process notifications
// tickers notifications
client.on('ticker', function(evt){
    console.log(`\n=== Got '${evt.pair}' 'ticker' event from '${evt.exchange}' === `);
    console.log(evt);
});


examples/subscribe_to_multiple_exchanges.js
"use strict";
const WebSocketClient = require('../lib/client');

/*
 * Below example assumes that your gateway is listening for WS on 127.0.0.1:8001
 */

const uri = 'ws://127.0.0.1:8001'
const client = new WebSocketClient(uri);

//-- Connection event handlers
// socket is connected but not ready to receive subscription yet
client.on('connected', function(data){
    console.log(`We connected/reconnected : ${JSON.stringify(data)}`);
});
// connection to gateway was lost (reconnection will be automatic)
client.on('disconnected', function(data){
    console.log(`We've been disconnected : ${JSON.stringify(data)}`);
});
// temporary connection error, retry will be automatic
client.on('connectionError', function(data){
    console.log(`Connection error : ${JSON.stringify(data)}`);
});
// this is were we can start subscriptions
client.on('ready', function(data){
    console.log(`Ready to subscribe : ${JSON.stringify(data)}`);

    // subscribe to USDT-BTC & USDT-ETH tickers on Poloniex
    client.subscribeToTickers('poloniex', ['USDT-BTC','USDT-ETH']);

    // subscribe to USDT-NEO order book on Bittrex
    client.subscribeToOrderBooks('bittrex', ['USDT-NEO']);

    // subscribe to BTC-NEO trades on Binance
    client.subscribeToTrades('binance', ['BTC-NEO']);

    // after 30s, ask to resync order book
    setTimeout(function(){
        client.resyncOrderBooks('bittrex', ['USDT-NEO']);
    }, 30000);
    // after 60s, unsubscribe from all exchanges
    setTimeout(function(){
        console.log("--- Unsubscribing from all exchanges ---");
        client.unsubscribe();
    }, 60000);
});

//-- Process notifications
// tickers notifications
client.on('ticker', function(evt){
    console.log(`\n=== Got '${evt.pair}' 'ticker' event from '${evt.exchange}' === `);
    console.log(evt);
});

// order books notifications
client.on('orderBook', function(evt){
    console.log(`\n=== Got '${evt.pair}' 'orderBook' event from '${evt.exchange}' === `);
    // just display buy/sell size
    let obj = {
        exchange:evt.exchange,
        pair:evt.pair,
        cseq:evt.cseq,
        buySize:evt.data.buy.length,
        sellSize:evt.data.sell.length
    }
    console.log(obj);
});
client.on('orderBookUpdate', function(evt){
    console.log(`\n=== Got '${evt.pair}' 'orderBookUpdate' event from '${evt.exchange}' === `);
    // just display buy/sell size
    let obj = {
        exchange:evt.exchange,
        pair:evt.pair,
        cseq:evt.cseq,
        buySize:evt.data.buy.length,
        sellSize:evt.data.sell.length
    }
    console.log(obj);
});

// trades notifications
client.on('trades', function(evt){
    console.log(`\n=== Got '${evt.pair}' 'trades' event from '${evt.exchange}' === `);
    // just display how many trades we're received
    let obj = {
        exchange:evt.exchange,
        pair:evt.pair,
        tradesSize:evt.data.length
    }
    console.log(evt);
});
